src/iop/useless.c

```diff
@@ -232,60 +290,185 @@ void cleanup_global(dt_iop_module_so_t *module)
-static void scale_callback(GtkWidget *w, dt_iop_module_t *self)
+static void extra_callback(GtkWidget *w, dt_iop_module_t *self)
 {
   // this is important to avoid cycles!
   if(darktable.gui->reset) return;
+
   dt_iop_useless_params_t *p = (dt_iop_useless_params_t *)self->params;
-  p->checker_scale = dt_bauhaus_slider_get(w);
+
+  float extra = dt_bauhaus_slider_get(w);
+
+  // setting a widget value will trigger a callback that will update params.
+  // if this is not desirable (because it might result in a cycle) then use
+  // ++darktable.gui->reset;
+  dt_bauhaus_slider_set_soft(g->factor, p->factor + extra);
+  // (and reverse with --darktable.gui->reset;)
+
+  // if any params updated directly, not via a callback, then
   // let core know of the changes
   dt_dev_add_history_item(darktable.develop, self, TRUE);
 }
 
-static void factor_callback(GtkWidget *w, dt_iop_module_t *self)
+/** optional gui callbacks. */
+void gui_changed(dt_iop_module_t *self, GtkWidget *w, void *previous)
 {
-  // this is important to avoid cycles!
-  if(darktable.gui->reset) return;
+  // if defined, this gets called when any of the introspection based widgets
+  // (created with dt_bauhaus_..._from_params) are changed.
+  // the updated value from the widget is already set in params.
+  // any additional side-effects can be achieved here.
   dt_iop_useless_params_t *p = (dt_iop_useless_params_t *)self->params;
-  p->factor = dt_bauhaus_slider_get(w);
-  // let core know of the changes
+  dt_iop_useless_gui_data_t *g = (dt_iop_useless_gui_data_t *)self->gui_data;
+
+  // test which widget was changed. if allowing w == NULL, this can be called
+  // from gui_update, so that gui configuration adjustments only need to be
+  // dealt with here.
+  if(!w || w == g->method)
+  {
+    gtk_widget_set_visible(g->check, p->method == DT_USELESS_SECOND);
+  }
+
+  // widget configurations that don't depend any any current params values should
+  // go in reload_defaults (if they depend on the picture) or gui_init.
+}
+
+void color_picker_apply(dt_iop_module_t *self, GtkWidget *picker, dt_dev_pixelpipe_iop_t *piece)
+{
+  dt_iop_useless_params_t *p = (dt_iop_useless_params_t *)self->params;
+  dt_iop_useless_gui_data_t *g = (dt_iop_useless_gui_data_t *)self->gui_data;
+
+  // this automatically gets called when any of the color pickers set up with
+  // dt_color_picker_new in gui_init is used. If there is more than one,
+  // check which one is active first.
+  if(picker == g->factor)
+  {
+    p->factor = self->picked_color[1];
+  }
+
   dt_dev_add_history_item(darktable.develop, self, TRUE);
+  dt_control_queue_redraw_widget(self->widget);
 }
 
-/** gui callbacks, these are needed. */
+/** gui setup and update, these are needed. */
 void gui_update(dt_iop_module_t *self)
 {
-  // let gui slider match current parameters:
+  // this gets called when switching to darkroom, with each picture change or when
```

Comment:
There are a couple of places where you say 'picture'. Pedantic I know but 'image' is probably the preferred term.

Reply:
thanks! I was vaguely aware I wasn't being consistent. searched-and-replaced!